### DISCUSSION TITLE
deploying via container image

### DISCUSSION THREAD

**Discussion by tcurdt** [caprover/caprover#1434] [General]
My CI/CD builds an image and pushes it into a private container registry (over at github).

How can I tell caprover via CI/CD (github action) that there is a new image to deploy - and ideally which version?

A token and webook seems ideal. Then the pipeline would only have to make a HTTP request and caprover could pull the new image from the registry. But so far I don't see that is a possible option.

How is this deployment path supposed to work?

**Comment by githubsaturn** [comment]
https://caprover.com/docs/ci-cd-integration.html

```
caprover deploy --caproverUrl $CAPROVER_URL --caproverPassword $CAPROVER_PASSWORD --caproverApp $CAPROVER_APP --imageName $CONTAINER_FULL_IMAGE_NAME_WITH_TAG
```

**Reply by tcurdt** [reply]
Thanks for the pointer. It sounds like creating an app token seems to be a better option though.

```
caprover deploy --appToken <YOUR_APP_TOKEN_HERE> --caproverUrl https://captain.domain.com --imageName YOUR_IMAGE_NAME --appName YOUR_APP_NAME
```

What does `caprover deploy` do under the hood? An authenticated HTTP API call to the caprover server I presume?

**Reply by githubsaturn** [reply]
Yes.

**Reply by tcurdt** [reply]
So in theory it should also be possible to even just use `curl` to transmit the data?

**Reply by githubsaturn** [reply]
Yes, it's essentially just two http calls. One to retrieve the auth token, the other one to deploy. But I recommend using the official method as it's more straightforward. You can use CLI docker image so that no footprint is left behind:
```
docker run caprover/cli-caprover:v2.1.1 caprover deploy --caproverUrl $CAPROVER_URL --caproverPassword $CAPROVER_PASSWORD --caproverApp $CAPROVER_APP --imageName $CONTAINER_FULL_IMAGE_NAME_WITH_TAG
  ```

**Reply by cbudtz** [reply]
This is sort of an undocumented feature. --imageName is not mentioned in the docs - can you supply more parameters? Like login information for docker hub private repos?

**Reply by githubsaturn** [reply]
@maietta  - CapRover itself runs in a container, hence it doesn't have access to the host's docker login creds.

To use a private repo, you need add the private registry credentials under cluster > add registry

Make sure to "disable default push" otherwise, by default, any new images built on CapRover will be also pushed to your private registry.

**Reply by maietta** [reply]
Thanks for the clarification. I removed my post as it was irrelevant to the conversation.
